### ISSUE TITLE
Swift interface for Objective C header AnyObject!

### ISSUE THREAD

**Issue by danielbowden** [realm/jazzy#609]
I've just documented an Objective C framework and noticed the Swift interfaces in the Jazzy output docs for these Objective C headers to be different to what I was expecting. Most data types are listed as `AnyObject!`

From this PR jpsim/SourceKitten#95 and realm/jazzy#530 it seems like this is supported so I'm thinking I might be missing some modern ObjC syntax to have the headers be parsed nicely with Swift?

For example the following method:

``` objc
- (instancetype)initWithClientID:(NSString *)clientID clientSecret:(NSString *)clientSecret collection:(NSString *)collection;
```

Renders in my jazzy output docs as 

``` objc
- (instancetype)initWithClientID:(NSString *)clientID
                    clientSecret:(NSString *)clientSecret
                      collection:(NSString *)collection;
```

``` swift
init!(clientID: AnyObject!, clientSecret: AnyObject!, collection: AnyObject!)
```

Or this method:

``` objc
- (void)myMethodName:(NSDictionary*)payload local:(BOOL)local;
```

``` objc
- (void)myMethodName:(NSDictionary *)payload
                       local:(BOOL)local;
```

``` swift
func myMethodName(payload: AnyObject!, local: AnyObject!)
```

I have seen this works in the **Realm-cocoa** docs for example:

``` objc
- (RLMResults *)sortedResultsUsingProperty:(NSString *)property ascending:(BOOL)ascending;
```

``` objc
- (nonnull RLMResults<RLMObjectType> *)
sortedResultsUsingProperty:(nonnull NSString *)property
                 ascending:(BOOL)ascending;
```

``` swift
func sortedResultsUsingProperty(property: String, ascending: Bool) -> RLMResults
```

I see that for the Realm docs the jazzy command line arguments include a `--swift-version` which I have tried. Am I missing something else?

``` sh
    jazzy \
      ${objc} \
      --swift-version 2.2 \
      --clean \
      --author Realm \
      --author_url https://realm.io \
      --github_url https://github.com/realm/realm-cocoa \
      --github-file-prefix https://github.com/realm/realm-cocoa/tree/v${version} \
      --module-version ${version} \
      --xcodebuild-arguments ${xcodebuild_arguments} \
      --module ${module} \
      --root-url https://realm.io/docs/${language}/${version}/api/ \
      --output docs/${language}_output \
      --head "$(cat docs/custom_head.html)"
```

**Comment by alexmx** [comment]
I'm experiencing the same issue with jazzy for Obj-C projects. I've tried on versions: 0.6.2 and 0.7.0.

Configuration: `jazzy --objc --umbrella-header "MyLib/MyLib.h" --framework-root "." --sdk iphonesimulator`

The doc set is generated successfully without any errors or warnings.

But parameter types are not translated properly for swift interface:
```objective-c
OBJECTIVE-C

- (nonnull instancetype)updateWithErrorCode:(nullable NSString *)errorCode
                                 errorTitle:(nullable NSString *)errorTitle
                               errorMessage:(nonnull NSString *)errorMessage;

SWIFT
func updateWithErrorCode(errorCode: AnyObject!, errorTitle: AnyObject!, errorMessage: AnyObject!) -> Self
```

**Comment by jpsim** [comment]
What happens when you generate the Swift interface for your projects in Xcode @danielbowden @alexmx? Jazzy uses the same mechanism, so if the interface generated by Xcode is the same as what jazzy is rendering, then you should file a bug at bugs.swift.org to improve this interface generation in Swift/SourceKit itself.

**Comment by alexmx** [comment]
@jpsim Xcode generated interface looks fine:

```objective-c
- (nonnull instancetype)updateWithErrorCode:(nullable NSString *)errorCode
                                 errorTitle:(nullable NSString *)errorTitle
                               errorMessage:(nonnull NSString *)errorMessage;
```

```swift
public func updateWithErrorCode(errorCode: String?, errorTitle: String?, errorMessage: String) -> Self
```

Jazzy generates:

```objective-c
OBJECTIVE-C
- (nonnull instancetype)updateWithErrorCode:(nullable NSString *)errorCode
                                 errorTitle:(nullable NSString *)errorTitle
                               errorMessage:(nonnull NSString *)errorMessage;
SWIFT
func updateWithErrorCode(errorCode: AnyObject!, errorTitle: AnyObject!, errorMessage: AnyObject!) -> Self
```

I observed that only `Foundation` types were translated wrong for my project, so I have fixed the issue by explicitly importing `Foundation` in the source files for which the generated interface was wrong. 
However I don't need this import as I'm importing the `UIKit` which exposes the foundation classes.

So Jazzy generates wrong Swift interface if there is no explicit import of `Foundation` in the interface files where the `Foundation` types are used, even though this import is not needed.

**Comment by jpsim** [comment]
@alexmx if you can share a sample project and exact steps to follow in Xcode that can generate a Swift interface for an Objective-C project that's different than the one jazzy generates, I can look into this more, reverse engineering what Xcode does and identifying what we may be doing differently.

**Comment by alexmx** [comment]
@jpsim Sure.

I've made a sample project where I'm able to reproduce the issue with `UIKit` types.

[TestProj.zip](https://github.com/realm/jazzy/files/388010/TestProj.zip)

Inside the root folder is the script `generate-docs.sh` which invokes `jazzy`.

This is a screenshot from my Xcode where you can see the Swift generated interface.
![screen shot 2016-07-28 at 9 31 28 am](https://cloud.githubusercontent.com/assets/1628901/17206908/f08a616a-54a9-11e6-87e7-699987608e57.png)

This is the Jazzy generated interface:
![screen shot 2016-07-28 at 10 00 36 am](https://cloud.githubusercontent.com/assets/1628901/17206961/32c2e23c-54aa-11e6-86c6-e825774a4f3a.png)

**Comment by danielbowden** [comment]
Thanks for continuing this @alexmx, I've been caught up

**Comment by JanC** [comment]
I have the exact same issue. It looks as if all types from UIKit were unknown. Foundation or custom objects are recognised correctly.

**Comment by jpsim** [comment]
Thanks for the sample project @alexmx! That helps a lot. Although it will certainly be tricky to replicate that functionality.

When I make Xcode generate the Swift interface to your `BestClass.h` header, Xcode kicks off 57 separate requests to SourceKit, in which the order and timing of these requests very much matters.

The first 2 requests for generating the interface return something similar to what Jazzy is producing:

```swift
open class func showMyAwesomeController(_ controller: Any!, presenter: Any!)
```

However, once Xcode figures out more of the compiler arguments involved, it can construct a more complete interface request:

```swift
open class func showMyAwesomeController(_ controller: UIViewController, presenter: UIViewController?)
```

I've uploaded the whole SourceKit log as a Gist [here](https://gist.github.com/jpsim/fb91816148ae08b3caf250f4d954b8c2) in case you want to help me reverse engineer this exquisite Rube Goldberg Machine of a process and wrap it up into SourceKitten's [DocCommand](https://github.com/jpsim/SourceKitten/blob/master/Source/sourcekitten/DocCommand.swift). 😄

**Comment by sfyz** [comment]
I think I'm running into a similar issue... If I just create a Single View Controller project in Objective-C and add this method to, say, the ViewController class:

`- (void)doSomethingWithBool:(BOOL)somebool withString:(NSString*)somestring withInt:(int)someint withColor:(UIColor*)somecolor;`

I'll see this Swift syntax in the documentation:

`func doSomething(withBool somebool: Any!, withString somestring: Any!, with someint: Int32, withColor somecolor: Any!)`

Thanks!

**Comment by jpsim** [comment]
Yes, that's the same issue. See https://github.com/realm/jazzy/issues/609#issuecomment-263036449

**Comment by michalzelinka** [comment]
Similar issue here:

```objc
- (void)placesForQuery:(TKPlacesQuery *)query
	completion:(void (^)(NSArray<TKPlace *>  * _Nullable places, NSError * _Nullable error))completion;

- (void)detailedPlaceWithID:(NSString *)placeID
	completion:(void (^)(TKPlace * _Nullable place, NSError * _Nullable error))completion;
```

gets this output in its interface counterpart:

```swift
open func places(for query: TKPlacesQuery, completion: @escaping ([TKPlace]?, Error?) -> Swift.Void)

open func detailedPlace(withID placeID: String, completion: @escaping (TKPlace?, Error?) -> Swift.Void)
```

but Jazzy puts out:

```swift
func places(forQuery query: Any!, completion: @escaping ([Any]?, Error?) -> Void)

func detailedPlace(withID placeID: String, completion: @escaping (UnsafeMutablePointer
```

First method fails in types and `forQuery:` instead `for:`, the second one scrambles the block type completely.

**Comment by johnfairh** [comment]
Going to close this as 'working as designed'.

Docs updated to explain the problem and how to fix it.

I wrote above about possibly changing the tools to compile the 'umbrella' to avoid needing workarounds.  This didn't go anywhere:
* The sourcekit API to get a Swift version of a C header is specifically for that header, does not follow imports
* The 'right' sourcekit API is the one to get a Swift version of a C _module_ interface, but our tools do not understand C modules and realistically they are very unlikely to ever learn
* I tried a couple of experiments using the preprocessor to create a single monster header from the umbrella and then use that, but couldn't get anything working robustly.
